crates/dbsp/src/operator/dynamic/aggregate/min.rs

```diff
@@ -140,3 +140,94 @@ where
         accumulator
     }
 }
+
+/// Given a tuple Tup1<(Option<K>, V)>, this aggregator that returns
+/// the smallest V for the smallest value K that is not None.  This is
+/// useful for implementing the SQL ARG_MIN operator.  Notice that the
+/// ARG_MIN(a, b) function returns the smallest a for the smallest b
+/// in the collection (it compares first on b); however, the compiler
+/// swaps the arguments when using this aggregator, so type Option<K>
+/// below is the type of b, and V is the type of A.  (This aggregator
+/// is used only when b is nullable, otherwise the regular Min
+/// aggregator is fine.)
+#[derive(Clone)]
+pub struct ArgMinSome;
+
+#[derive(Clone)]
+pub struct ArgMinSomeSemigroup<K, V>(PhantomData<(K, V)>);
+
+impl<K, V> Semigroup<Tup1<(Option<K>, V)>> for ArgMinSomeSemigroup<K, V>
+where
+    K: DBData,
+    V: DBData,
+{
+    fn combine(left: &Tup1<(Option<K>, V)>, right: &Tup1<(Option<K>, V)>) -> Tup1<(Option<K>, V)> {
+        match (&left.0 .0, &right.0 .0) {
+            (None, None) => min(left, right).clone(),
+            (Some(_), None) => left.clone(),
+            (None, Some(_)) => right.clone(),
+            (Some(_), Some(_)) => min(left, right).clone(),
+        }
+    }
+}
+
+impl<V, K, T, R> Aggregator<Tup1<(Option<K>, V)>, T, R> for ArgMinSome
+where
+    K: DBData,
+    V: DBData,
+    T: Timestamp,
+    R: DBWeight + MonoidValue,
+{
+    type Accumulator = Tup1<(Option<K>, V)>;
+    type Output = Tup1<V>;
+    type Semigroup = ArgMinSomeSemigroup<K, V>;
+
+    fn aggregate<VTrait, RTrait>(
+        &self,
+        cursor: &mut dyn Cursor<VTrait, DynUnit, T, RTrait>,
+    ) -> Option<Self::Accumulator>
+    where
+        VTrait: DataTrait + ?Sized,
+        RTrait: WeightTrait + ?Sized,
+        Tup1<(Option<K>, V)>: Erase<VTrait>,
+        R: Erase<RTrait>,
+    {
+        // Result will be None if the cursor points to an empty collection
+        let mut result = None;
+
+        while cursor.key_valid() {
+            // FIXME: This could be more succinct if we had an `Add<&Self, Output = Self>`
+            // bound on `R`
+            let mut weight: R = HasZero::zero();
+            cursor.map_times(&mut |_, w| {
+                weight.add_assign_by_ref(unsafe { w.downcast() });
+            });
+            if !weight.is_zero() {
+                let current = unsafe { cursor.key().downcast::<Tup1<(Option<K>, V)>>() };
+                match current.0 {
+                    (Some(_), _) => return Some(current.clone()),
+                    (None, _) => {
+                        if result.is_none() {
+                            // remember first value, this may be the final result if
+                            // all tuples have None in the first field
+                            result = Some(current.clone());
```

Comment:
I'm still not clear about this. I thought that if all tuples have None is the first field, the aggregate should return None. Isn't that the semantics in SQL?

Reply:
There is no ARG_MAX in standard SQL; this is an extension that exists in some dialects.
The semantics is not well-defined, and it is nondeterministic in some implementations.
In other implementations it works as you say.
But I prefer our (new) semantics (which we haven't really documented yet):
- it is deterministic
- it matches MIN/MAX
- it always returns a tuple that exists in the input collection

Reply:
> it is deterministic

we don't have a choice there really.

> it matches MIN/MAX

Doesn't the behavior I described match MIN: if the collection contains NULLs only, it returns NULL.

>  But I prefer our (new) semantics (which we haven't really documented yet)

This I don't fully understand. If there is no standard, we can at least follow the precedent. 

Anyhow, this seems good otherwise.